### PR TITLE
feat(workspace): add new resource to manage desktop name rules

### PR DESCRIPTION
**Pull request by wuzhuanhong**
**What this PR does / why we need it**:

Support a new resource to manage desktop name rule.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add a new resource to manage desktop name rule.
2. add related document and acceptance test.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST=./huaweicloud/services/acceptance/workspace TESTARGS='-run TestAccDesktopNameRule_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/workspace -v -run TestAccDesktopNameRule_basic -timeout 360m -parallel 4
=== RUN   TestAccDesktopNameRule_basic
=== PAUSE TestAccDesktopNameRule_basic
=== CONT  TestAccDesktopNameRule_basic
--- PASS: TestAccDesktopNameRule_basic (39.31s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/workspace 39.361s
```

### CODE REVIEW

File: docs/resources/workspace_desktop_name_rule.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+* `region` - (Optional, String, ForceNew) Specifies the region in which to create the resource.
+  If omitted, the provider-level region will be used.
+  Changing this creates a new resource.
+
+* `name` - (Required, String) Specifies the name of the rule.
+  The name can contain `1` to `30` characters, only digits, letters and underscores (_) are allowed.
+  The name must start with a letter or an underscore.
+  
+* `name_prefix` - (Required, String) Specifies the prefix of desktop name.
+  The format is as follows: `$DomainUser$`, `$DomainUser$`, `xx$DomainUser$`, `$DomainUser$xx`, `xx$DomainUser$xx`, `xx`.
```
Comment (Lance52259) [inline docs/resources/workspace_desktop_name_rule.md]:
```suggestion
  The format are as follows: `$DomainUser$`, `$DomainUser$`, `xx$DomainUser$`, `$DomainUser$xx`, `xx$DomainUser$xx`, `xx`.
```

File: docs/resources/workspace_desktop_name_rule.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  If the `digit_number` parameter is set to `2`, the valid value is range from `1` to `99`, and so on.
+
+* `single_domain_user_increment` - (Required, Int) Specifies whether to increment by single user name.
+  The valid values are as follows:
+  + **1**: Increment by single user name.
+  + **0**: Increment by tenant.
+
+  e.g. Assume that there are three user A, B, C. Allocate two desktops to user A, one to user B, and one to user C.
+  The `digit_number` parameter is set to `2`, the `start_number` parameter is set to `1`.
+  If the `single_domain_user_increment` is set to `1`, the desktop names are as follows: `A01`, `A02`, `B01`, `C02`.
```
Comment (Lance52259) [inline docs/resources/workspace_desktop_name_rule.md]:
```suggestion
  If the `single_domain_user_increment` is set to `1`, the desktop names are as follows: `A01`, `A02`, `B01`, `C01`.
```

File: docs/resources/workspace_desktop_name_rule.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  If the `single_domain_user_increment` is set to `0`, the desktop names are as follows: `A01`, `A02`, `B03`, `C04`.
+
+  -> If the `name_prefix` parameter not contain `$DomainUser$`, the `single_domain_user_increment` value must set to `0`.
+
+* `is_default_policy` - (Optional, Bool) Specifies whether to set as default rule. The default value is **false**.
+  
+  -> Only one default policy is allowed, if the parameter is set to `true`, the parameter of other data is changed to `false`.
```
Comment (Lance52259) [inline docs/resources/workspace_desktop_name_rule.md]:
```suggestion
  ~> Only one default policy is allowed, if the parameter is set to `true`, the parameter of other data is changed to `false`.
```
using warning and update the description of the change result.

File: docs/resources/workspace_desktop_name_rule.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
+---
+subcategory: "Workspace"
+---
+
+# huaweicloud_workspace_desktop_name_rule
+
+Manages a desktop name rule resource within HuaweiCloud.
```
Comment (Lance52259) [inline docs/resources/workspace_desktop_name_rule.md]:
```suggestion
Manages a desktop name rule resource within HuaweiCloud.
Using this resource to manage a name rule for the desktops creation within HuaweiCloud.
```

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		httpUrl = "v2/{project_id}/desktop-name-policies"
+	)
+
+	client, err := cfg.WorkspaceV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating Workspace v2 client: %s", err)
+	}
```
Comment (Lance52259) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go]:
```suggestion
	client, err := cfg.NewServiceClient(region)
	if err != nil {
		return diag.Errorf("error creating Workspace client: %s", err)
	}
```
The version info is never used in the createPath building phases

File: docs/resources/workspace_desktop_name_rule.md (hunk 1 above)
Comment (wuzhuanhong) [reply docs/resources/workspace_desktop_name_rule.md]: fixed

File: docs/resources/workspace_desktop_name_rule.md (hunk 2 above)
Comment (wuzhuanhong) [reply docs/resources/workspace_desktop_name_rule.md]: fixed

File: docs/resources/workspace_desktop_name_rule.md (hunk 3 above)
Comment (wuzhuanhong) [reply docs/resources/workspace_desktop_name_rule.md]: fixed

File: docs/resources/workspace_desktop_name_rule.md (hunk 4 above)
Comment (wuzhuanhong) [reply docs/resources/workspace_desktop_name_rule.md]: fixed

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go (hunk 5 above)
Comment (wuzhuanhong) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go]: fixed

File: huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	)
+
+	client, err := cfg.NewServiceClient("workspace", region)
+	if err != nil {
+		return diag.Errorf("error creating Workspace client: %s", err)
+	}
+
+	rule, err := GetDesktopNameRule(client, d.Id())
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "Workspace name rule")
```
Comment (Lance52259) [inline huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go]:
```suggestion
		return common.CheckDeletedDiag(d, err, "desktop name rule")
```
desktop or workspace? pls check the other places

Comment (wuzhuanhong) [reply huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop_name_rule.go]: fixed

**Pull request merged**